chore(docs) Docs for new CLI and Starter (#820)
electric-sql/electric: docs/api/cli.md

This will download all migrations from the database, bundle them with your application, and generate the Electric client.
Note that you can use an optional `--watch` flag to automatically re-generate the client on every database migration.

#### Options

The `generate` command accepts a number of arguments:

```shell

> npx electric-sql generate [--service <url>] [--proxy <url>] [--out <path>] [--watch [<pollingInterval>]]

Reviewer: This synopsis does not match the list of options below.

|----------------------|--------------|------------------------------------------|
| <span className="no-wrap">`--service`</span><br/><span className="no-wrap">`-s`</span><br/>`ELECTRIC_SERVICE` |`<url>` | Provides the url to connect to the [Electric sync service](./service.md).<br /> Defaults to `http://localhost:5133`. |
| <span className="no-wrap">`--proxy`</span><br/><span className="no-wrap">`-p`</span><br/>`ELECTRIC_PROXY` | `<url>` | Provides the url to connect to Electric's database proxy.<br /> Defaults to <span class="break-all">`postgresql://prisma:proxy_password@localhost:65432/electric`</span>. |
| <span className="no-wrap">`--client-path`</span><br /><span className="no-wrap">`-o`</span><br />`ELECTRIC_CLIENT_PATH` | `<path>` | Specifies the output location for the generated client.<br /> Defaults to `./src/generated/client` |
| <span className="no-wrap">`--watch`</span><br /><span className="no-wrap">`-w`</span> | `<pollingInterval>` | Run the generator in watch mode. Accepts an optional polling interval (in milliseconds) which defines how often to poll Electric for database migrations.<br /> The default polling interval is 1000ms. |

> | <span className="no-wrap">`--with-migrations`</span> | `<command>` | Specify a command to run migrations against a blank postgres in order to create a client. [See details below](#generate---with-migrations) |

Reviewer: ```suggestion
| <span className="no-wrap">`--with-migrations`</span> | `<command>` | Specify a command to run migrations against an ephemeral database in order to create a client. [See details below](#generate---with-migrations) |
```

| <span className="no-wrap">`--module-resolution`</span> | `<command>` | The module resolution used for the project. The generated client will be compatible with this resolution scheme, [see notes below](#electric_module_resolution). |

:::caution
Note that the username in the `--proxy` URL **must** be `prisma`.

This is to activate the proxy mode that uses Prisma tooling for schema introspection. It does not mean that your Postgres database actually needs a `prisma` user.
:::

For a full list of arguments run `npx electric-sql help generate` or see the [environment variables](#environment-variables) below.

> #### Local-only-first mode  

Reviewer: There's a fair amount of duplication between this section and the one right below it:
![image](https://github.com/electric-sql/electric/assets/207748/49afee3e-0719-46fe-9878-b74ed014ee30)

Maybe leave high-level description here, in the `Local-only-first mode` intro and leave the technical details about the underlying commands for the next section?

This will download all migrations from the database, bundle them with your application, and generate the Electric client.
Note that you can use an optional `--watch` flag to automatically re-generate the client on every database migration.

#### Options

The `generate` command accepts a number of arguments:

```shell
npx electric-sql generate [--service <url>] [--proxy <url>] [--out <path>] [--watch [<pollingInterval>]]
```

All arguments are optional. The principal ones are described below:

| Argument or Env var  | Value        | Description                              |
|----------------------|--------------|------------------------------------------|
| <span className="no-wrap">`--service`</span><br/><span className="no-wrap">`-s`</span><br/>`ELECTRIC_SERVICE` |`<url>` | Provides the url to connect to the Electric sync service.<br /> Defaults to `http://localhost:5133`. |

> | <span className="no-wrap">`--proxy`</span><br/><span className="no-wrap">`-p`</span><br/>`ELECTRIC_PROXY` | `<url>` | Provides the url to connect to Electric's database proxy.<br /> Defaults to <span class="break-all">`postgresql://prisma:proxy_password@localhost:65432/electric`</span>. |

Reviewer 2: Note that it's essential to document the need for the username to be `prisma`. As per the clashing docs I just merged.

| <span className="no-wrap">`--client-path`</span><br /><span className="no-wrap">`-o`</span><br />`ELECTRIC_CLIENT_PATH` | `<path>` | Specifies the output location for the generated client. Defaults to `./src/generated/client` |
| <span className="no-wrap">`--watch`</span><br /><span className="no-wrap">`-w`</span> | `<pollingInterval>` | Run the generator in watch mode. Accepts an optional polling interval (in milliseconds) which defines how often to poll Electric for database migrations.<br /> The default polling interval is 1000ms. |

> | <span className="no-wrap">`--with-migrations`</span> | `<command>` | Specify a command to generate migrations. With this option the work flow is:<br /> 1. Start new ElectricSQL and PostgreSQL containers<br /> 2. Run the provided migrations command<br /> 3. Generate the client<br /> 4. Stop and remove the containers.<br /> See the [`with-config`](#with-config) command for details. |

Reviewer 2: I see the link to the with-config section here, which has the usage example of with-migrations. However, I think maybe it deserves more prominence here where the generate command is.

Ideally I'd like to see a section explaining the local-only-first flow.

| <span className="no-wrap">`--client-path`</span><br /><span className="no-wrap">`-o`</span><br />`ELECTRIC_CLIENT_PATH` | `<path>` | Specifies the output location for the generated client.<br /> Defaults to `./src/generated/client` |
| <span className="no-wrap">`--watch`</span><br /><span className="no-wrap">`-w`</span> | `<pollingInterval>` | Run the generator in watch mode. Accepts an optional polling interval (in milliseconds) which defines how often to poll Electric for database migrations.<br /> The default polling interval is 1000ms. |
| <span className="no-wrap">`--with-migrations`</span> | `<command>` | Specify a command to run migrations against a blank postgres in order to create a client. [See details below](#generate---with-migrations) |

:::caution
Note that the username in the `--proxy` URL **must** be `prisma`.

This is to activate the proxy mode that uses Prisma tooling for schema introspection. It does not mean that your Postgres database actually needs a `prisma` user.
:::

For a full list of arguments run `npx electric-sql help generate` or see the [environment variables](#environment-variables) below.

#### `generate --with-migrations`

> Normally the `generate` command expects to be run against an Electric and Postgres where you already have a full schema installed. However you may want to work in a *"local only first"* way, generating a client directly from either SQL migrations or using a migration tool such as Prisma, without having to continually run a Postgres server and Electric sync service.

Reviewer 2: This is great but I think it can be tightened up one more notch by focusing the message on "running the backend or not". I.e.:

```md
### Local-only-first mode

Normally, when you develop or deploy apps with Electric, you want to run the backend services (Postgres and Electric) in order to sync data. As a result, it's natural that the `generate` command expects you to be running the backend services during your build step.

However, it's also quite common in development or with an early version of an app, not to need data sync enabled yet. Either because you're just working on the interface or because your app starts off local-only and you plan to enable sync later. This makes the overhead of running (and potentially deploying) the backend services quite high, given that you're running them just to support the generate command in your build step.

As a result, we provide a special `--with-migrations` mode for the generate command that allows you to generate the type safe client just from your migrations, without having to run the backend services yourself. Specifically what the `--with-migrations` option does is it tells the generate command to spin up a temporary Postgres and Electric itself, apply your migrations from scratch, generate the type safe client and then tear Electric and Postgres down. All of which happens automatically in the background for you.

We call this approach "local-only-first", in the sense that it allows you to develop local-only and then progressively enable sync later on, as and when you want to.

<then the command examples etc. from below>

As you can see from the steps above, the backend is started, your migrations are applied, the type-safe client is generated and then everything is torn down and cleaned up for you. Allowing you to develop local-only-first and then run the backend services only when you actually want to enable sync.
```

Author: Good call, I've push an update with that

| <span className="no-wrap">`--image`</span><br />`ELECTRIC_IMAGE` | `<image>` | The Docker image to use for Electric. |
| <span className="no-wrap">`--postgresql-image`</span><br />`ELECTRIC_POSTGRESQL_IMAGE` | `<image>` | The Docker image to use for the PostgreSQL database. |

For a full list of arguments run `npx electric-sql help start` or see the [environment variables](#environment-variables) below.

### `stop`

Stop the development ElectricSQL sync service that was started with the [`start`](#start) command, and any optional PostgreSQL.

```shell
npx electric-sql stop [--remove]
```

#### Options

The `start` command accepts a single argument:

| Argument             | Description                              |
|----------------------|------------------------------------------|
| <span className="no-wrap">`--remove`</span> | Remove the containers and volumes from Docker.

### `status`

Show status of the ElectricSQL sync service docker containers that were started with the [`start`](#start) command.

```shell
npx electric-sql status
```

### `psql`

> Connect with psql to the ElectricSQL PostgreSQL proxy.

Reviewer 2: It's the `Migrations proxy`. And I would say: Start an interactive PSQL session with Postgres, connecting via the Electric [Migrations proxy](...).

```shell
npx electric-sql psql [--proxy <url>]
```

#### Options

> The `psql` command accepts a number of arguments for specifying the Postgres Proxy to connect to, the main one being:

Reviewer 2: Migrations proxy.

| Argument or Env var  | Value        | Description                              |
|----------------------|--------------|------------------------------------------|
| <span className="no-wrap">`--proxy`</span><br /><span className="no-wrap">`-p`</span><br />`ELECTRIC_PROXY` | `<url>` | URL of the Electric service's PostgreSQL proxy.

For a full list of arguments run `npx electric-sql help psql` or see the [environment variables](#environment-variables) below.

### `configure-ports`

> Configure the ports used by the ElectricSQL sync service.

Reviewer 2: Electric sync service.

This starts an interactive session where you are asked to specify the ports you would like to use. These are then written to your `.env.local` file, and used for the other commands.

```shell
npx electric-sql configure-ports
```

### `show-config`

Print out the full configuration that the CLI is using based on any env variable, or `.env` files.

```shell
npx electric-sql show-config
```

### `with-config`

> This command allows you to run a subcommand substituting arguments configured with Electric environment variables. It also makes all Electric configuration environment variables available to the subcommand.

Reviewer 2: I don't understand from this what this does. And what kind of command would you run? Can the description be clearer and can you provide a real example?

Reviewer 2: Sorry, I see you have further below. I leave the commend here though because I didn't read ahead that far before being confused. You may want to bring the example higher up.

```

One of the main use cases for this command is running migrations against your Electric Postgres Proxy using the configuration specified locally. Many of our examples use the [@databases/pg-migrations](https://www.atdatabases.org/docs/pg-migrations) tool for migrations, and we run the migrations with:

```shell
npx electric-sql with-config "npx pg-migrations apply --database {{ELECTRIC_PROXY}} --directory ./db/migrations"
```

We have this configured as a "script" in our `package.json` for each example:

```js
// package.json
{
  //...
  "scripts": {
    "db:migrate": "npx electric-sql with-config \"npx pg-migrations apply --database {{ELECTRIC_PROXY}} --directory ./db/migrations\"",
    // ...
  }
}
```

The [`generate`](#generate) command takes an optional `--with-migrations` command that also takes the same argument form as the `with-config` command. With this you can run:

```shell
npx electric-sql generate --with-migrations "npx pg-migrations apply --database {{ELECTRIC_PROXY}} --directory ./db/migrations"
```

This will perform these tasks, allowing you to generate a client directly from a set of migrations:

> 1. Start new ElectricSQL and PostgreSQL containers

Reviewer 2: As per comment in the generate command above, it's great to show this usage but I think we could highlight the benefit and principle more. I.e.: this allows you to develop with Electric without running the backend and without sync. That's a big deal but the lede isn't clear here.

```

#### Options

The `proxy-tunnel` command accepts a number of arguments:

```shell
npx electric-sql proxy-tunnel [--service <url>] [--local-port <port>]
```

All arguments are optional and are described below:

| Argument or Env var  | Value        | Description                              |
|----------------------|--------------|------------------------------------------|
| <span className="no-wrap">`--service`</span><br/>`ELECTRIC_SERVICE` | `<url>` | Provides the url to connect to the Electric sync service. If not provided it uses the url set in the `ELECTRIC_URL` environment variable. If that variable is not set, it resorts to the default url which is `http://localhost:5133`. |
| <span className="no-wrap">`--local-port`</span> | `<port>` | The local port to bind to; this will be forwarded to the Electric sync service, and defaults to `65432`. |

### `start`

Starts an Electric sync service (using Docker) for development, along with an optional Postgres database.

By default it will launch a sync service that is compatible with the client version you have installed.

> All [environment variables for configuring the Electric sync service](./service.md#configuration-options)  are passed through to the service if they are found prefixed with `ELECTRIC_`.

Reviewer: That anchor is no longer there.

```suggestion
All [environment variables for configuring the Electric sync service](./service.md)  are passed through to the service if they are found prefixed with `ELECTRIC_`.
```

Reviewer: > if they are found prefixed with `ELECTRIC_`

I'm not sure what this means. Only variables prefixed with `ELECTRIC` are passed through? Does the developer need to add the `ELECTRIC_` prefix to the variables documented on the `service.md` page?

To start an Electric sync service, along with Postgres, fully configured for development run:

```shell
npx electric-sql start --with-postgres
```

#### Options

The `start` command accepts a number of arguments:

```shell
npx electric-sql start [options]
```

All arguments are optional. The principal ones are described below:

| Argument or Env var  | Value        | Description                              |
|----------------------|--------------|------------------------------------------|
| <span className="no-wrap">`--with-postgres`</span> |  | Start a Postgres database along with Electric.
| <span className="no-wrap">`--detach`</span> |  | Run in the background instead of printing logs to the console.

> | <span className="no-wrap">`--database-url`</span><br/><span className="no-wrap">`-db`</span><br/>`ELECTRIC_DATABASE_URL` | `<url>` | PostgreSQL connection URL for the database. |

Reviewer: It's not clear what database this is referring to. Is this option mutually exclusive with `--with-postgres`? If not, how do the two interact?

| <span className="no-wrap">`--http-port`</span><br />`ELECTRIC_HTTP_PORT` | `<port>` | The local port to run the sync service on. Defaults to 5133 |
| <span className="no-wrap">`--pg-proxy-port`</span><br />`ELECTRIC_PG_PROXY_PORT` | `<port>` | The local port to bind the Postgres Proxy port to. Defaults to 65432 |
| <span className="no-wrap">`--image`</span><br />`ELECTRIC_IMAGE` | `<image>` | The Docker image to use for Electric. |
| <span className="no-wrap">`--postgresql-image`</span><br />`ELECTRIC_POSTGRESQL_IMAGE` | `<image>` | The Docker image to use for the PostgreSQL database. |

For a full list of arguments run `npx electric-sql help start` or see the [environment variables](#environment-variables) below.

### `stop`

Stop the development Electric sync service that was started with the [`start`](#start) command, and any optional PostgreSQL.

```shell
npx electric-sql stop [--remove]
```

#### Options

> The `start` command accepts a single argument:

Reviewer: ```suggestion
The `stop` command accepts a single argument:
```

| Argument             | Description                              |
|----------------------|------------------------------------------|

> | <span className="no-wrap">`--remove`</span> | Remove the containers and volumes from Docker.

Reviewer: ```suggestion
| <span className="no-wrap">`--remove`</span> | Remove the Docker containers and volumes previously created by `start`.
```

### `status`

Show status of the Electric sync service docker containers that were started with the [`start`](#start) command.

```shell
npx electric-sql status
```

### `psql`

> Start an interactive PSQL session with Postgres, connecting via the Electric [Migrations proxy](../usage/data-modelling/migrations.md#migrations-proxy).

Reviewer: Does this require `psql` to be installed? If yes, let's mention that.

```shell
npx electric-sql psql [--proxy <url>]
```

#### Options

> The `psql` command accepts a number of arguments for specifying the Migrations proxy to connect to, the main one being:

Reviewer: I find this structure very confusing: this sentence mentions "a number of arguments", the table below contains a single one. One has to go through the full list of environment variables and see in the description of each var which command it applies to. That's inconvenient. Moreover, it's not clear how one should map the env var name to a command-line flag.

| Argument or Env var  | Value        | Description                              |
|----------------------|--------------|------------------------------------------|
| <span className="no-wrap">`--proxy`</span><br /><span className="no-wrap">`-p`</span><br />`ELECTRIC_PROXY` | `<url>` | URL of the Electric service's PostgreSQL proxy.

For a full list of arguments run `npx electric-sql help psql` or see the [environment variables](#environment-variables) below.

### `show-config`

Print out the full configuration that the CLI is using based on any env variable, or `.env` files.

```shell
npx electric-sql show-config
```

### `with-config`

This command allows you to run a subcommand substituting arguments configured with Electric environment variables. It also makes all Electric configuration environment variables available to the subcommand.

> One of the main use cases for this command is running migrations against your Electric Postgres Proxy using the configuration specified locally. 

Reviewer: Throughout this page, it's called `Migration proxy`, `Migrations proxy` and now `Electric Postgres Proxy`. I think we need to standardize on a single name so as not to confuse people.

It takes a single string argument, which is the command to run in string form, and substitutes configuration values inside double braces:

```shell
npx electric-sql with-config "run-migration --db {{ELECTRIC_PROXY}}"
```

Many of our examples use the [@databases/pg-migrations](https://www.atdatabases.org/docs/pg-migrations) tool for migrations, and we run the migrations with:

```shell
npx electric-sql with-config "npx pg-migrations apply --database {{ELECTRIC_PROXY}} --directory ./db/migrations"
```

We have this configured as a "script" in our `package.json` for each example:

```js
// package.json
{
  //...
  "scripts": {

>     "db:migrate": "npx electric-sql with-config \"npx pg-migrations apply --database {{ELECTRIC_PROXY}} --directory ./db/migrations\"",

Reviewer: How about this to improve readability?

```suggestion
    "db:migrate": "npx electric-sql with-config 'npx pg-migrations apply --database {{ELECTRIC_PROXY}} --directory ./db/migrations'",
```

    // ...
  }
}
```

### `help`

Display the help for a command.

```shell
npx electric-sql help <command>
```

> ## Environment Variables

Reviewer: I can't seem to guess the ordering criterion used for variables in this section. Let's order them alphabetically? Or at least by logically related groups. Currently, `ELECTRIC_SERVICE` is followed by two unrelated vars before `ELECTRIC_SERVICE_HOST`.  Then, after 7 other env vars comes `ELECTRIC_HTTP_PORT`.

> In addition to those specified below, all standard Electric sync service environment variables are available for setting prefixed with `ELECTRIC_`. These are then passed though to the sync service started with the [`start`](#start) command. See a [full list of environment variables here](./service.md#configuration-options).

Reviewer: ```suggestion
In addition to those specified below, all standard Electric sync service environment variables are available for setting prefixed with `ELECTRIC_`. These are then passed though to the sync service started with the [`start`](#start) command. See a [full list of environment variables here](./service.md).
```

import EnvVarConfig from '@site/src/components/EnvVarConfig'

#### `ELECTRIC_SERVICE`

<EnvVarConfig
    name="ELECTRIC_SERVICE"
    defaultValue="http://{ELECTRIC_SERVICE_HOST}:{ELECTRIC_HTTP_PORT}"

>     example="http://electric.mydomain.com"

Reviewer: ```suggestion
    example="https://electric.mydomain.com"
```

>
  URL of the Electric service.

  Used by the [`generate`](#generate) command.

  With all defaults, it evaluates to `http://localhost:5133`
</EnvVarConfig>

This `ELECTRIC_SERVICE` variable is perfect for configuring the Electric client in your project. 

```typescript
import { schema } from './generated/client'

const conn = await ElectricDatabase.init(scopedDbName)
const electric = await electrify(conn, schema, {
  url: import.meta.env.ELECTRIC_SERVICE
  // ...
})
```

> #### `ELECTRIC_PROXY`

Reviewer: Perhaps `ELECTRIC_PG_PROXY` is a more suitable name, to be consistent with its constituent `PG_PROXY_PORT` and `PG_PROXY_PASSWORD` variables?
